File: python/attention.py
# SPDX-License-Identifier: CC-BY-NC-4.0
# Copyright (c) 2025-2026 fumi-engineer

"""Multi-Query Attention with RoPE for MoE Transformer.

Implements scaled dot-product attention with:
  - Multi-Query Attention (MQA): n_kv_heads << n_heads, KV heads shared
  - Rotary Position Embedding (RoPE) with optional NTK-aware scaling
  - Causal (autoregressive) masking

Attention: scores = Q @ K^T / sqrt(d_k), weights = softmax(scores + mask), output = weights @ V
"""

from __future__ import annotations

import math

import numpy as np

from .config import Config
from .layers import Linear
from .tensor import Tensor


class MQAttention:
    """Multi-Query Attention with RoPE.

    MQA uses a single (or few) KV head(s) shared across all Q heads.
    This reduces the KV cache size by n_heads/n_kv_heads during inference
    while preserving most of the model quality of full MHA.
    """

    def __init__(self, config: Config):
        """Initialize attention layer."""
        self.config = config
        self.hidden_dim = config.hidden_dim
        self.n_heads = config.n_heads
        self.n_kv_heads = config.n_kv_heads
        self.head_dim = config.head_dim
        # Scaling factor: 1/sqrt(d_k) prevents dot products from growing
        # too large, which would push softmax into saturated regions.
        self.scale = 1.0 / math.sqrt(config.head_dim)

        # Projections — Q has n_heads, K/V have n_kv_heads (MQA)
        self.w_q = Linear(config.hidden_dim, config.n_heads * config.head_dim)
        self.w_k = Linear(config.hidden_dim, config.n_kv_heads * config.head_dim)
        self.w_v = Linear(config.hidden_dim, config.n_kv_heads * config.head_dim)
        self.w_o = Linear(config.n_heads * config.head_dim, config.hidden_dim)

        # RoPE parameters
        self.rope_base = config.rope_base
        self.rope_alpha = config.rope_alpha

        # Causal mask cache — avoids re-allocating a [seq, seq] mask every forward call.
        # Invalidated when seq_len changes.
        self._cached_mask: np.ndarray | None = None
        self._cached_mask_len: int = 0

        # KV head expansion indices — cached to avoid recomputation per forward call
        self._kv_indices = np.arange(self.n_heads) % self.n_kv_heads

    def _compute_rope_freqs(self, seq_len: int) -> np.ndarray:
        """Compute RoPE frequency bands.

        RoPE frequencies: freq_i = 1 / (base^(2i/d))  for i in [0, d/2)

        With NTK scaling (alpha > 1):
          base' = base * alpha^(d/(d-2))
        This stretches the frequency spectrum to support longer sequences
        without retraining.
        """
        base = self.rope_base
        if self.rope_alpha > 1.0:
            # NTK-aware scaling: base' = base * alpha^(d/(d-2))
            base = self.rope_base * (
                self.rope_alpha ** (self.head_dim / (self.head_dim - 2))
            )

        i = np.arange(self.head_dim // 2, dtype=np.float32)
        return (1.0 / (base ** (2 * i / self.head_dim))).astype(np.float32)

    def _apply_rope(
        self, q: np.ndarray, k: np.ndarray, seq_len: int
    ) -> tuple[np.ndarray, np.ndarray]:
        """Apply Rotary Position Embedding to Q and K.

        RoPE rotation for each pair of dimensions (2i, 2i+1):
          [x_even, x_odd] -> [x_even*cos(theta) - x_odd*sin(theta),
                               x_even*sin(theta) + x_odd*cos(theta)]
        where theta = position * freq_i.

        This encodes absolute position through relative-position-aware
        dot products: q_m . k_n depends only on (m - n).
        """
        freqs = self._compute_rope_freqs(seq_len)

        # angles[pos, i] = pos * freq_i   shape: [seq_len, head_dim/2]
        positions = np.arange(seq_len, dtype=np.float32)
        angles = np.outer(positions, freqs)

        # Broadcast to [1, seq_len, 1, half_dim] for batched Q/K
        cos_vals = np.cos(angles)[None, :, None, :]
        sin_vals = np.sin(angles)[None, :, None, :]

        # Split even/odd dims: q is [batch, seq, heads, head_dim]
        q_even = q[:, :, :, 0::2]
        q_odd = q[:, :, :, 1::2]
        k_even = k[:, :, :, 0::2]
        k_odd = k[:, :, :, 1::2]

        # Apply 2D rotation per pair
        q_out = np.empty_like(q)
        q_out[:, :, :, 0::2] = q_even * cos_vals - q_odd * sin_vals
        q_out[:, :, :, 1::2] = q_even * sin_vals + q_odd * cos_vals

        k_out = np.empty_like(k)
        k_out[:, :, :, 0::2] = k_even * cos_vals - k_odd * sin_vals
        k_out[:, :, :, 1::2] = k_even * sin_vals + k_odd * cos_vals

        return q_out, k_out

    def forward(self, x: Tensor) -> Tensor:
        """Multi-query attention forward pass.

        Full formula:
          Q, K, V = x @ W_q, x @ W_k, x @ W_v
          Q, K    = RoPE(Q), RoPE(K)
          scores  = (Q @ K^T) / sqrt(d_k) + causal_mask
          weights = softmax(scores)
          output  = (weights @ V) @ W_o

        Args:
            x: Input [batch, seq_len, hidden_dim]

        Returns:
            Output [batch, seq_len, hidden_dim]
        """
        batch, seq_len, _ = x.shape

        # Project Q, K, V
        q = self.w_q.forward(x)  # [batch, seq_len, n_heads * head_dim]
        k = self.w_k.forward(x)  # [batch, seq_len, n_kv_heads * head_dim]
        v = self.w_v.forward(x)  # [batch, seq_len, n_kv_heads * head_dim]

        # Reshape to [batch, seq_len, n_heads, head_dim] using raw numpy (no Tensor overhead)
        q_data = q.data.reshape(batch, seq_len, self.n_heads, self.head_dim)
        k_data = k.data.reshape(batch, seq_len, self.n_kv_heads, self.head_dim)
        v_data = v.data.reshape(batch, seq_len, self.n_kv_heads, self.head_dim)

        # Apply RoPE to Q and K
        q_data, k_data = self._apply_rope(q_data, k_data, seq_len)

        # GQA/MQA head expansion: map each Q head to its corresponding KV head.
        # With n_kv_heads=1 (MQA), all Q heads share the same single KV head.
        # Uses cached _kv_indices to avoid recomputation per forward call.
        k_expanded = k_data[:, :, self._kv_indices, :]
        v_expanded = v_data[:, :, self._kv_indices, :]

        # Transpose to [batch, heads, seq, dim] for batched matmul
        q_t = np.transpose(q_data, (0, 2, 1, 3))
        k_t = np.transpose(k_expanded, (0, 2, 1, 3))
        v_t = np.transpose(v_expanded, (0, 2, 1, 3))

        # Attention scores: (Q @ K^T) / sqrt(d_k)
        # np.matmul dispatches to BLAS for the batched [heads, seq, dim] @ [heads, dim, seq] matmul
        scores = np.matmul(q_t, k_t.transpose(0, 1, 3, 2)) * self.scale

        # Causal mask: upper-triangular -inf prevents attending to future tokens.
        # Cached to avoid re-allocation when seq_len is unchanged between calls.
        if self._cached_mask_len != seq_len:
            self._cached_mask = np.where(
                np.triu(np.ones((seq_len, seq_len), dtype=np.float32), k=1) > 0,
                float('-inf'),
                0.0,
            )
            self._cached_mask_len = seq_len
        scores += self._cached_mask

        # Numerically stable softmax: subtract max before exp to prevent overflow
        # Softmax: p_i = exp(s_i - max(s)) / sum_j(exp(s_j - max(s)))
        shifted = scores - np.max(scores, axis=-1, keepdims=True)
        exp_vals = np.exp(shifted)
        attn_weights = exp_vals / np.sum(exp_vals, axis=-1, keepdims=True)

        # Store for backward pass
        self.last_attn_weights = attn_weights
        self.last_q = q_data
        self.last_k = k_data
        self.last_v = v_data
        self.last_input = x

        # Weighted sum of values: weights @ V
        # np.matmul dispatches to BLAS for [heads, seq, seq] @ [heads, seq, dim]
        output = np.matmul(attn_weights, v_t)
        # Transpose [batch, heads, seq, dim] -> [batch, seq, heads, dim] and flatten heads
        output = np.ascontiguousarray(np.transpose(output, (0, 2, 1, 3)))
        output = output.reshape(batch, seq_len, self.n_heads * self.head_dim)
        return self.w_o.forward(Tensor.from_numpy(output))

    def backward(self, grad_output: Tensor) -> Tensor:
        """Backward pass through attention layer.

        Propagates gradients through: W_o -> attention -> W_q, W_k, W_v.
        Uses stored attention weights and Q/K/V from the forward pass.

        The attention backward computes:
          grad w.r.t. V:  attn_weights^T @ grad_attn_out
          grad w.r.t. attn_weights: grad_attn_out @ V^T
          grad w.r.t. scores: softmax jacobian (simplified as grad_weights * weights * (1 - weights))
          grad w.r.t. Q: grad_scores @ K
          grad w.r.t. K: grad_scores^T @ Q
        Then propagates through the projection layers for weight gradient accumulation.
        """
        batch, seq_len, _ = grad_output.shape

        # 1. Backward through W_o
        grad_o_input = self.w_o.backward(grad_output)

        # Reshape grad to [batch, seq_len, n_heads, head_dim]
        grad_o_reshaped = grad_o_input.data.reshape(batch, seq_len, self.n_heads, self.head_dim)
        # Transpose to [batch, n_heads, seq_len, head_dim]
        grad_attn_out = np.transpose(grad_o_reshaped, (0, 2, 1, 3))

        # 2. Backward through attention: output = attn_weights @ V
        attn_weights = self.last_attn_weights  # [batch, n_heads, seq, seq]
        v_data = self.last_v  # [batch, seq, n_kv_heads, head_dim]
        v_expanded = v_data[:, :, self._kv_indices, :]  # [batch, seq, n_heads, head_dim]
        v_t = np.transpose(v_expanded, (0, 2, 1, 3))  # [batch, n_heads, seq, head_dim]

        # grad_V = attn_weights^T @ grad_attn_out  -> [batch, n_heads, seq, head_dim]
        grad_v_t = np.matmul(attn_weights.transpose(0, 1, 3, 2), grad_attn_out)

        # grad_attn_weights = grad_attn_out @ V^T  -> [batch, n_heads, seq, seq]
        grad_attn_w = np.matmul(grad_attn_out, v_t.transpose(0, 1, 3, 2))

        # 3. Backward through softmax: simplified Jacobian
        # d_scores = attn_weights * (grad_attn_w - sum(grad_attn_w * attn_weights, axis=-1, keepdims))
        sum_term = np.sum(grad_attn_w * attn_weights, axis=-1, keepdims=True)
        grad_scores = attn_weights * (grad_attn_w - sum_term)

        # Apply scale factor
        grad_scores *= self.scale

        # 4. Backward through Q @ K^T
        q_data = self.last_q  # [batch, seq, n_heads, head_dim]
        k_data = self.last_k  # [batch, seq, n_kv_heads, head_dim]
        k_expanded = k_data[:, :, self._kv_indices, :]
        q_t = np.transpose(q_data, (0, 2, 1, 3))  # [batch, n_heads, seq, head_dim]
        k_t = np.transpose(k_expanded, (0, 2, 1, 3))

        # grad_Q = grad_scores @ K  -> [batch, n_heads, seq, head_dim]
        grad_q_t = np.matmul(grad_scores, k_t)
        # grad_K = grad_scores^T @ Q  -> [batch, n_heads, seq, head_dim]
        grad_k_t = np.matmul(grad_scores.transpose(0, 1, 3, 2), q_t)

        # Transpose back to [batch, seq, n_heads, head_dim]
        grad_q = np.transpose(grad_q_t, (0, 2, 1, 3)).reshape(batch, seq_len, -1)
        grad_k = np.transpose(grad_k_t, (0, 2, 1, 3))
        grad_v = np.transpose(grad_v_t, (0, 2, 1, 3))

        # Reduce expanded heads back to kv_heads by summing
        # Head mapping is interleaved: [0,1,...,kv-1, 0,1,...,kv-1, ...]
        # Reshape to [batch, seq, heads_per_kv, n_kv_heads, head_dim] and sum
        heads_per_kv = self.n_heads // self.n_kv_heads
        grad_k_reduced = grad_k.reshape(
            batch, seq_len, heads_per_kv, self.n_kv_heads, self.head_dim
        ).sum(axis=2)
        grad_v_reduced = grad_v.reshape(
            batch, seq_len, heads_per_kv, self.n_kv_heads, self.head_dim
        ).sum(axis=2)

        grad_k_flat = grad_k_reduced.reshape(batch, seq_len, -1)
        grad_v_flat = grad_v_reduced.reshape(batch, seq_len, -1)

        # 5. Backward through projection layers (accumulates weight gradients)
        grad_q_tensor = Tensor.from_numpy(grad_q)
        grad_k_tensor = Tensor.from_numpy(grad_k_flat)
        grad_v_tensor = Tensor.from_numpy(grad_v_flat)

        # Set _last_input for Q/K/V projections (they all used the same input x)
        self.w_q._last_input = self.last_input
        self.w_k._last_input = self.last_input
        self.w_v._last_input = self.last_input

        grad_x_q = self.w_q.backward(grad_q_tensor)
        grad_x_k = self.w_k.backward(grad_k_tensor)
        grad_x_v = self.w_v.backward(grad_v_tensor)

        # Sum gradients from all projection paths
        grad_x = grad_x_q.data + grad_x_k.data + grad_x_v.data
        return Tensor.from_numpy(grad_x)

    def parameters(self) -> list[Tensor]:
        """Return all parameters (Q, K, V, O projections)."""
        return (
            self.w_q.parameters()
            + self.w_k.parameters()
            + self.w_v.parameters()
            + self.w_o.parameters()
        )

File: go/attention.go
// SPDX-License-Identifier: CC-BY-NC-SA-4.0
// Copyright (c) 2025-2026 fumi-engineer

package nn

// MQAttention implements Multi-Query Attention (MQA) with Rotary Position
// Embeddings (RoPE). In MQA, query heads outnumber key/value heads
// (nHeads > nKVHeads); each KV head is shared by (nHeads / nKVHeads) query heads,
// reducing KV cache memory in inference.
//
// Full attention equation:
//   scores = (Q @ K^T) / sqrt(d_k)
//   weights = softmax(scores + causal_mask)
//   output = weights @ V
//
// RoPE applies position-dependent rotation to Q and K before the dot product:
//   [x0, x1] -> [x0*cos(theta) - x1*sin(theta), x0*sin(theta) + x1*cos(theta)]
//   where theta_i = pos / base^(2i/d_head)
type MQAttention struct {
	wQ, wK, wV, wO            *Linear
	nHeads, nKVHeads, headDim int
	hiddenDim                 int
	scale                     float32   // 1 / sqrt(head_dim)
	freqs                     []float32 // precomputed RoPE frequency bands
	scoresBuf                 []float32 // reusable buffer for attention scores
	attnOutBuf                []float32 // reusable buffer for attention output
	// Cached from forward pass for backward
	lastInput       *Tensor     // input to attention
	lastQ           *Tensor     // Q after RoPE [batch, seq, nHeads, headDim]
	lastK           *Tensor     // K after RoPE [batch, seq, nKVHeads, headDim]
	lastV           *Tensor     // V [batch, seq, nKVHeads, headDim]
	lastAttnWeights []float32   // softmax weights [batch * nHeads * seq * seq]
	lastBatch       int
	lastSeqLen      int
}

// NewMQAttention creates a Multi-Query Attention layer.
//
// RoPE frequency computation:
//   base' = base * alpha^(d_head / (d_head - 2))   (NTK-aware scaling if alpha > 1)
//   freq_i = 1 / base'^(2i / d_head)               for i in [0, d_head/2)
func NewMQAttention(hiddenDim, nHeads, nKVHeads, headDim int, ropeBase, ropeAlpha float32) *MQAttention {
	base := ropeBase
	if ropeAlpha > 1.0 {
		// NTK-aware RoPE scaling: increase the base to extend context length
		// without fine-tuning. The exponent d/(d-2) is the NTK-aware formula.
		base = ropeBase * PowF32(ropeAlpha, float32(headDim)/float32(headDim-2))
	}
	freqs := make([]float32, headDim/2)
	for i := range freqs {
		freqs[i] = 1.0 / PowF32(base, float32(2*i)/float32(headDim))
	}

	return &MQAttention{
		wQ:     NewLinear(hiddenDim, nHeads*headDim, false),
		wK:     NewLinear(hiddenDim, nKVHeads*headDim, false),
		wV:     NewLinear(hiddenDim, nKVHeads*headDim, false),
		wO:     NewLinear(nHeads*headDim, hiddenDim, false),
		nHeads: nHeads, nKVHeads: nKVHeads, headDim: headDim,
		hiddenDim: hiddenDim,
		scale:     1.0 / SqrtF32(float32(headDim)),
		freqs:     freqs,
	}
}

// Forward computes Multi-Query Attention with causal masking.
//
// Steps:
//  1. Project: Q = W_Q @ x, K = W_K @ x, V = W_V @ x
//  2. Reshape to [batch, seq, heads, head_dim]
//  3. Apply RoPE to Q and K
//  4. Compute scores = Q @ K^T / sqrt(d_k) with causal mask (upper triangle = -inf)
//  5. Softmax over valid positions (up to and including current position)
//  6. Weighted sum: output = softmax_weights @ V
//  7. Reshape and project output through W_O
//
// The causal mask is implemented by only iterating ki <= qi in the score loop,
// filling future positions with NegInf before softmax.
func (a *MQAttention) Forward(input *Tensor) *Tensor {
	dims := input.Shape().DimsRef()
	batch, seqLen := dims[0], dims[1]
	a.lastInput = input
	a.lastBatch = batch
	a.lastSeqLen = seqLen

	q := a.wQ.Forward(input).Reshape(NewShape(batch, seqLen, a.nHeads, a.headDim))
	k := a.wK.Forward(input).Reshape(NewShape(batch, seqLen, a.nKVHeads, a.headDim))
	v := a.wV.Forward(input).Reshape(NewShape(batch, seqLen, a.nKVHeads, a.headDim))

	a.applyRoPE(q.DataPtr(), k.DataPtr(), batch, seqLen, a.nHeads, a.nKVHeads)

	// Cache Q, K, V after RoPE for backward pass
	a.lastQ = q.Clone()
	a.lastK = k.Clone()
	a.lastV = v.Clone()

	// Reuse attention output buffer to avoid allocation per forward pass.
	outLen := batch * seqLen * a.nHeads * a.headDim
	if cap(a.attnOutBuf) >= outLen {
		a.attnOutBuf = a.attnOutBuf[:outLen]
		for i := range a.attnOutBuf {
			a.attnOutBuf[i] = 0
		}
	} else {
		a.attnOutBuf = make([]float32, outLen)
	}
	output := FromSliceNoCopy(a.attnOutBuf, NewShape(batch, seqLen, a.nHeads, a.headDim))
	outData, qData, kData, vData := output.DataPtr(), q.DataPtr(), k.DataPtr(), v.DataPtr()

	// Allocate attention weights storage: [batch * nHeads * seqLen * seqLen]
	attnWeightsLen := batch * a.nHeads * seqLen * seqLen
	if len(a.lastAttnWeights) < attnWeightsLen {
		a.lastAttnWeights = make([]float32, attnWeightsLen)
	} else {
		a.lastAttnWeights = a.lastAttnWeights[:attnWeightsLen]
		for i := range a.lastAttnWeights {
			a.lastAttnWeights[i] = 0
		}
	}

	// Reuse scores buffer to avoid allocation per forward pass.
	scoresLen := seqLen * seqLen
	if cap(a.scoresBuf) >= scoresLen {
		a.scoresBuf = a.scoresBuf[:scoresLen]
	} else {
		a.scoresBuf = make([]float32, scoresLen)
	}
	scores := a.scoresBuf
	for b := 0; b < batch; b++ {
		for h := 0; h < a.nHeads; h++ {
			// Multi-Query: map query head h to KV head (h % nKVHeads)
			kvH := h % a.nKVHeads

			// Compute Q @ K^T / sqrt(d_k) with causal mask.
			for qi := 0; qi < seqLen; qi++ {
				qOff := ((b*seqLen+qi)*a.nHeads + h) * a.headDim
				qRow := qData[qOff : qOff+a.headDim]
				sRow := scores[qi*seqLen : (qi+1)*seqLen]

				for ki := 0; ki <= qi; ki++ {
					kOff := ((b*seqLen+ki)*a.nKVHeads + kvH) * a.headDim
					kRow := kData[kOff : kOff+a.headDim]
					dot := float32(0)
					for d := range qRow {
						dot += qRow[d] * kRow[d]
					}
					sRow[ki] = dot * a.scale
				}
				// Causal mask: future positions get -inf so softmax zeroes them.
				for ki := qi + 1; ki < seqLen; ki++ {
					sRow[ki] = NegInf
				}
			}

			// Softmax each row up to the causal boundary.
			// Zero masked positions so backward BLAS sees 0, not -inf.
			for qi := 0; qi < seqLen; qi++ {
				softmaxInPlace(scores[qi*seqLen : qi*seqLen+qi+1])
				for ki := qi + 1; ki < seqLen; ki++ {
					scores[qi*seqLen+ki] = 0
				}
			}

			// Save attention weights for backward pass
			// Layout: [batch, nHeads, seqLen, seqLen]
			awOff := (b*a.nHeads + h) * seqLen * seqLen
			copy(a.lastAttnWeights[awOff:awOff+seqLen*seqLen], scores[:seqLen*seqLen])

			// Weighted sum: output = attention_weights @ V
			for qi := 0; qi < seqLen; qi++ {
				outOff := ((b*seqLen+qi)*a.nHeads + h) * a.headDim
				oRow := outData[outOff : outOff+a.headDim]
				for ki := 0; ki <= qi; ki++ {
					w := scores[qi*seqLen+ki]
					vOff := ((b*seqLen+ki)*a.nKVHeads + kvH) * a.headDim
					vRow := vData[vOff : vOff+a.headDim]
					for d := range oRow {
						oRow[d] += w * vRow[d]
					}
				}
			}
		}
	}

	// Concatenate heads: [batch, seq, nHeads, headDim] -> [batch, seq, nHeads*headDim]
	output = output.Reshape(NewShape(batch, seqLen, a.nHeads*a.headDim))
	return a.wO.Forward(output)
}

// Backward computes the full attention backward pass.
// Propagates gradients through: W_o -> attention (V, weights, softmax, scores) -> W_q, W_k, W_v.
//
// Uses per-head BLAS sgemm calls with strided access to [batch, seq, nHeads, headDim] tensors.
func (a *MQAttention) Backward(gradOutput *Tensor) *Tensor {
	batch, seqLen := a.lastBatch, a.lastSeqLen

	// 1. Backward through W_o: gradOInput shape [batch, seq, nHeads*headDim]
	gradOInput := a.wO.Backward(gradOutput)
	goData := gradOInput.DataPtr()

	// Q, K, V data from cached tensors [batch, seq, heads, headDim]
	qData := a.lastQ.DataPtr()
	kData := a.lastK.DataPtr()
	vData := a.lastV.DataPtr()

	// Allocate gradient tensors for Q, K, V (zero-initialized)
	gradQ := make([]float32, batch*seqLen*a.nHeads*a.headDim)
	gradK := make([]float32, batch*seqLen*a.nKVHeads*a.headDim)
	gradV := make([]float32, batch*seqLen*a.nKVHeads*a.headDim)

	// Scratch buffer for grad_scores per head [seqLen * seqLen]
	gradScores := make([]float32, seqLen*seqLen)

	hd := a.headDim
	qStride := a.nHeads * hd
	kvStride := a.nKVHeads * hd

	for b := 0; b < batch; b++ {
		for h := 0; h < a.nHeads; h++ {
			kvH := h % a.nKVHeads
			awOff := (b*a.nHeads + h) * seqLen * seqLen

			// Base offsets for this (batch, head) into the [batch, seq, heads, headDim] tensors
			qBase := b*seqLen*qStride + h*hd
			kvBase := b*seqLen*kvStride + kvH*hd
			goBase := qBase
			gqBase := qBase
			gkBase := kvBase
			gvBase := kvBase

			// 2. grad_V += W^T @ dO
			// W is [seqLen, seqLen], dO is [seqLen, headDim]
			// gradV[seqLen, headDim] += W^T[seqLen, seqLen] @ dO[seqLen, headDim]
			sgemmRaw(true, false,
				seqLen, hd, seqLen,
				1.0,
				a.lastAttnWeights[awOff:], seqLen,
				goData[goBase:], qStride,
				1.0, // beta=1.0: accumulate into gradV (multiple Q heads may share this KV head)
				gradV[gvBase:], kvStride)

			// 3. grad_W = dO @ V^T -> gradScores
			// dO[seqLen, headDim] @ V^T[headDim, seqLen] = gradScores[seqLen, seqLen]
			sgemmRaw(false, true,
				seqLen, seqLen, hd,
				1.0,
				goData[goBase:], qStride,
				vData[kvBase:], kvStride,
				0.0, // beta=0.0: overwrite gradScores
				gradScores, seqLen)

			// 4. Softmax backward (element-wise)
			// gradScores = weights * (gradScores - sum(gradScores * weights))
			for qi := 0; qi < seqLen; qi++ {
				row := qi * seqLen
				sumTerm := float32(0)
				for ki := 0; ki <= qi; ki++ {
					sumTerm += gradScores[row+ki] * a.lastAttnWeights[awOff+row+ki]
				}
				for ki := 0; ki <= qi; ki++ {
					w := a.lastAttnWeights[awOff+row+ki]
					gradScores[row+ki] = w * (gradScores[row+ki] - sumTerm)
				}
				// Zero future positions (sgemm wrote values there from the full V^T multiply)
				for ki := qi + 1; ki < seqLen; ki++ {
					gradScores[row+ki] = 0
				}
			}

			// 5. grad_Q = scale * grad_scores @ K
			// gradScores[seqLen, seqLen] @ K[seqLen, headDim] = gradQ[seqLen, headDim]
			sgemmRaw(false, false,
				seqLen, hd, seqLen,
				a.scale,
				gradScores, seqLen,
				kData[kvBase:], kvStride,
				0.0, // beta=0.0: each Q head is unique, no accumulation needed
				gradQ[gqBase:], qStride)

			// 6. grad_K += scale * grad_scores^T @ Q
			// gradScores^T[seqLen, seqLen] @ Q[seqLen, headDim] = gradK[seqLen, headDim]
			sgemmRaw(true, false,
				seqLen, hd, seqLen,
				a.scale,
				gradScores, seqLen,
				qData[qBase:], qStride,
				1.0, // beta=1.0: accumulate into gradK (multiple Q heads may share this KV head)
				gradK[gkBase:], kvStride)
		}
	}

	// 7. Backward through Q, K, V projection layers
	// Flatten gradQ from [batch, seq, nHeads, headDim] to [batch, seq, nHeads*headDim]
	gradQTensor := FromSliceNoCopy(gradQ, NewShape(batch, seqLen, a.nHeads*a.headDim))
	gradKTensor := FromSliceNoCopy(gradK, NewShape(batch, seqLen, a.nKVHeads*a.headDim))
	gradVTensor := FromSliceNoCopy(gradV, NewShape(batch, seqLen, a.nKVHeads*a.headDim))

	// Set lastInput for Q/K/V projections (they all used the same input x)
	a.wQ.lastInput = a.lastInput
	a.wK.lastInput = a.lastInput
	a.wV.lastInput = a.lastInput

	gradXQ := a.wQ.Backward(gradQTensor)
	gradXK := a.wK.Backward(gradKTensor)
	gradXV := a.wV.Backward(gradVTensor)

	// Sum gradients from all projection paths
	return gradXQ.Add(gradXK).Add(gradXV)
}

// Parameters returns all projection weights: Q, K, V, and O.
func (a *MQAttention) Parameters() []*Tensor {
	return concatParams(
		a.wQ.Parameters(),
		a.wK.Parameters(),
		a.wV.Parameters(),
		a.wO.Parameters(),
	)
}

// applyRoPE applies Rotary Position Embeddings in-place to Q and K tensors.
//
// RoPE rotates consecutive pairs of dimensions by a position-dependent angle:
//   theta_i = position * freq_i
//   [x_{2i}, x_{2i+1}] -> [x_{2i}*cos(theta) - x_{2i+1}*sin(theta),
//                           x_{2i}*sin(theta) + x_{2i+1}*cos(theta)]
//
// This encodes relative position information directly into the Q/K vectors
// so that the dot product Q_i . K_j depends on (i - j), not absolute positions.
func (a *MQAttention) applyRoPE(qData, kData []float32, batch, seqLen, qHeads, kHeads int) {
	halfDim := a.headDim / 2
	rotate := func(data []float32, heads int, b, s int) {
		base := (b*seqLen + s) * heads * a.headDim
		pos := float32(s)
		for h := 0; h < heads; h++ {
			off := base + h*a.headDim
			row := data[off : off+a.headDim]
			for i := 0; i < halfDim; i++ {
				angle := pos * a.freqs[i]
				cos, sin := CosF32(angle), SinF32(angle)
				x0, x1 := row[2*i], row[2*i+1]
				row[2*i] = x0*cos - x1*sin
				row[2*i+1] = x0*sin + x1*cos
			}
		}
	}

	for b := 0; b < batch; b++ {
		for s := 0; s < seqLen; s++ {
			rotate(qData, qHeads, b, s)
			rotate(kData, kHeads, b, s)
		}
	}
}
